### PR TITLE
Filepicker: Add option to return additional details for selected file(s)

### PR DESCRIPTION
**Pull request by petermueller87**
Currently the Filepicker only returns a list of paths to the selected files.
It would be great to have an option to retrieve additional details about selected file(s), (mimetype, size, ..)

### CODE REVIEW

**Comment by karlitschek** [comment]
@petermueller87 Thanks a lot. When/How/Where is this useful?

**Comment by petermueller87** [comment]
@karlitschek I want to integrate ownCloud into my app.
A user should be able to choose files from his ownCloud drive to share them via WebRTC.
Only a list of files isn't really helpful. I need at least size and mimetype as additional information.

**Comment by karlitschek** [comment]
Interesting. @PVince81 @DeepDiver1975 What do you think?

**Comment by PVince81** [comment]
From what I remember the file picker is also using ajax/list.php which returns the required information.
This is mostly a JS API problem that the file picker code doesn't return that information.

**Comment by petermueller87** [comment]
How else could one easily retrieve the mimetype of given file paths?

File: core/js/oc-dialogs.js [hunk 1]
```diff
@@ -185,17 +187,28 @@ var OCdialogs = {
...
-						datapath = self.$filePicker.data('path');
-						datapath += '/' + self.$filelist.find('.filepicker_element_selected .filename').text();
+						if (multiselect === true) {
+							selected = [];
+							self.$filelist.find('.filepicker_element_selected .filename').each(function(index, element) {
+								selected.push(self.$filePicker.data('path') + '/' + $(element).text());
+							});
+						} else {
+							selected = self.$filePicker.data('path');
+							selected += '/' + self.$filelist.find('.filepicker_element_selected .filename').text();
```
Comment (PVince81) [inline core/js/oc-dialogs.js]: Is there no "data" property for the file name ? Using .text() is a bit dirty.

You might also want to use `OC.joinPaths(path1, path2)` to safely join path entries.

Comment (PVince81) [inline core/js/oc-dialogs.js]: Right, seems `.text()` was already used before.

**Comment by PVince81** [comment]
I agree with @petermueller87 that the file picker should also return more file information. Many of the APIs are/were much too restrictive.

@DeepDiver1975 what happened to the contribution message from the bot ?

**Comment by PVince81** [comment]
also I remember that several people have asked for this feature on IRC lately.

@petermueller87 how about returning a simpler array ?
Instead of:
```json
[{
   path: '/path/to/file.txt',
   details: {
       mimetype: 'text/plain',
       size: 123,
       ...
   }
},
{...}
]
```
make it flatter like:
```
[{
   selectedPath: '/path/to/file.txt',
   mimetype: 'text/plain',
   size: 123,
},
{...}
]
```
You can merge two arrays using `_.extend({selectedPath: '/path/to/file.txt'}, details)`.
I suggest to rename "path" to "selectedPath" in that case because sometimes the details already have a "path" attribute which is the local path, not full path.

**Comment by petermueller87** [comment]
Updated PR.
Now we properly check for `if (withDetails === undefined) {` to be consistent.

**Comment by petermueller87** [comment]
@PVince81 sure, good idea. Should I update the PR?

**Comment by PVince81** [comment]
>sure, good idea. Should I update the PR?

Please do.

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 what happened to the contribution message from the bot ?

Hmmm - good question .... let me check ...

**Comment by DeepDiver1975** [comment]
> Hmmm - good question .... let me check ...

No idea ... issue in the Jenkins plugin ... no the first time.... we should use something more reliable for this .... https://cla-assistant.io/

**Comment by petermueller87** [comment]
PR has been updated.

**Comment by PVince81** [comment]
- [x] TEST: no-detail file picker single select => returns single string
- [x] TEST: no-detail file picker multiple select => returns array of strings
- [x] TEST: withDetails file picker single select => returns array of details with single element
- [ ] TEST: withDetails file picker multiple select => returns array of details with multiple elements :no_entry_sign: FAIL

@petermueller87 multi-select only returns a single entry, but should return several detail entries:
filepicker-api1

**Comment by petermueller87** [comment]
@PVince81 I'm so sorry for this. Issue has been fixed.

**Comment by petermueller87** [comment]
BTW: How to actually select multiple files from the filepicker?

**Comment by PVince81** [comment]
Haha, good question... I'd expect checkboxes as in the regular file list but there don't seem to be any. This component is quite old and could use an update...

Anyway, to select multiple files use Ctrl+click. It is important to retest all combinations to make sure everything still works.

Normally I'd also ask for a Javascript unit test to cover this but as there is currently no good example for this it might be hard, so would postpone the JS unit test for another time.

**Comment by petermueller87** [comment]
Ctrl+click shows the context menu on Mac OS X, thus one cannot select multiple files on OS X.
Will file an issue for that.

**Comment by petermueller87** [comment]
#18228

**Comment by PVince81** [comment]
@petermueller87 any update ? Did it work selecting with the command key instead ? (not sure how it's done usually on mac)

**Comment by MorrisJobke** [comment]
We are past feature freeze for 8.2 -> I will move this to the 9.0 milestone.

**Comment by DeepDiver1975** [comment]
@PVince81 where are we going with this?
@petermueller87 are you still with us? :wink:

**Comment by PVince81** [comment]
We're past feature freeze. Either we close due to lack of inactivity this or move to 9.1 or backlog if someone wants to pick it up.

**Comment by PVince81** [comment]
Closing due to lack of activity

**Pull request closed without merging**